File: demos/pax-britannica/pax-britannica-android/assets/data/scripts/fighter_ai.lua
local v2 = require 'dokidoki.v2'

-- shot range
local shot_range = 200
-- try to stay this far away when you're out of ammo
local run_distance = 200

-- true when we've shot everything and want to make a distance, false means
-- we're approaching to attack
local running = false

local target = false
local on_screen = true

local function retarget ()
  target = game.targeting.get_nearest_of_type(self, 'bomber') or
           game.targeting.get_nearest_of_type(self, 'fighter') or
           game.targeting.get_nearest_of_type(self, 'frigate') or           
           game.targeting.get_nearest_of_type(self, 'factory')
end

-- basically, the behaviour is to go towards the target and attack it, then
-- when you're out of ammo run away until you've got your ammo back and have
-- gone a certain distance

function update()
  -- if we go from on to off screen, retarget
  local new_on_screen = game.targeting.on_screen(self.transform.pos)
  if (on_screen and not new_on_screen)
     or not target
     or target.dead
     or math.random() < 0.005 then
    retarget()
  end
  on_screen = new_on_screen

  if target then
    game.tracing.trace_line(self.transform.pos, target.transform.pos)
  end

  if target then
    local to_target = target.transform.pos - self.transform.pos
    local dist_squared = v2.sqrmag(to_target)

    if running then
      -- run away until you have full ammo and are far enough away

      local too_close = dist_squared < run_distance^2
      -- if you're too close to the target then turn away
      if too_close then
        self.ship.go_away(target.transform.pos, true)
      else
        self.ship.thrust()
      end

      if not self.fighter_shooting.is_empty() and not too_close then
        running = false
      end
    else
      -- go towards the target and attack!

      self.ship.go_towards(target.transform.pos, true)

      -- maybe shoot
      if self.fighter_shooting.is_ready_to_shoot() then
        local facing = self.transform.facing
        if dist_squared <= shot_range * shot_range
          and v2.dot(to_target, facing) > 0
          and v2.dot(to_target, facing)^2 > 0.97 * dist_squared then
          self.fighter_shooting.shoot()
        end
      end

      -- if out of shots then run away
      if self.fighter_shooting.is_empty() then
        running = true
      end
    end
  end
end

File: demos/pax-britannica/pax-britannica/src/de/swagner/paxbritannica/fighter/FighterAI.java
package de.swagner.paxbritannica.fighter;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;

import de.swagner.paxbritannica.Ship;
import de.swagner.paxbritannica.Targeting;

public class FighterAI {
	// shot range
	private float shot_range = 200;

	// try to stay this far away when you're out of ammo
	private float run_distance = 200;

	// true when we've shot everything and want to make a distance, false means
	// we're approaching to attack
	private boolean running = false;

	public Ship target;
	private boolean on_screen = true;
	
	//recycle vars
	Vector2 to_target = new Vector2();

	private Fighter fighter;

	public FighterAI(Fighter fighter) {
		this.fighter = fighter;
		retarget();
	}

	public void retarget() {
		target = Targeting.getNearestOfType(fighter, 1);
		if (target == null) {
			target = Targeting.getNearestOfType(fighter, 0);
		}
		if (target == null) {
			target = Targeting.getNearestOfType(fighter, 2);
		}
		if (target == null) {
			target = Targeting.getNearestOfType(fighter, 3);
		}
	}

	public void update() {
		// if we go from on to off screen, retarget
		boolean new_on_screen = Targeting.onScreen(fighter.collisionCenter);
		if (on_screen && !new_on_screen || target == null || !target.alive || MathUtils.random() < 0.005f) {
			retarget();
		}
		on_screen = new_on_screen;

		if (target != null) {
			to_target.set(target.collisionCenter.x - fighter.collisionCenter.x, target.collisionCenter.y - fighter.collisionCenter.y);
			float dist_squared = to_target.dot(to_target);

			if (running) {
				// run away until you have full ammo and are far enough away
				boolean too_close = dist_squared < Math.pow(run_distance, 2);
				// if you're too close to the target then turn away
				if (too_close) {
					fighter.goAway(target.collisionCenter, true);
				} else {
					fighter.thrust();
				}

				if (!fighter.isEmpty() && !too_close) {
					running = false;
				}
			} else {
				// go towards the target and attack!
				fighter.goTowards(target.collisionCenter, true);

				// maybe shoot
				if (fighter.isReadyToShoot()) {
					if (dist_squared <= shot_range * shot_range && to_target.dot(fighter.facing) > 0 && Math.pow(to_target.dot(fighter.facing), 2) > 0.97 * dist_squared) {
						fighter.shoot();
					}
				}

				// if out of shots then run away
				if (fighter.isEmpty()) {
					running = true;
				}
			}
		}
	}
}
